pkg/persistence/disk_persistence.go

```diff
@@ -65,6 +82,53 @@ func (ds *diskPersistence) Save(data []byte, dirName, fileName string) error {
 	return write(fmt.Sprintf("%s/%s/%s", dirPath, dirName, fileName), data)
 }
 
+func (ds *diskPersistence) Snapshot(data []byte, dirName, fileName string) error {
```

Comment:
Why implement this as a separate feature of disk persistence, instead of introducing it as a separate persistence handler?

My thinking here is: having it as a handler instead of something that must be explicitly called removes the possibility for any downstream code to forget to snapshot when it would be useful. My thought on what this would look like:

```go
	handle, err := persistence.NewDiskHandle(config.Storage.DataDir)
	if err != nil {
		return fmt.Errorf("failed while creating a storage disk handler: [%v]", err)
	}
	dataLogger := persistence.NewDataLogger(handle)
	persistence := persistence.NewEncryptedPersistence(
		dataLogger,
		config.Ethereum.Account.KeyFilePassword,
	)

```

Anything written to `persistence` would get written to `dataLogger` before being written to regular disk persistence. No downstream management, no nothing---if you write it to disk, it gets snapshotted. If you want a disk handle that doesn't snapshot for some reason, you can create that separately.

This also allows future modifications where, say, the data logger writes to disk but the disk handle is replaced by an `S3Handle` or vice versa.

Thoughts?

Reply:
I have made it as a separate feature because of the way how it will be used by the `keep-ecdsa` client. My reasoning was:

1. The client generates the signer 
2. Before submitting the public key on-chain, the client makes a signer snapshot but doesn't persist the signer in the regular storage yet. This is because the signer is still a "draft" until the public key gets submitted on-chain so the client can't put it in the regular storage.
3. Once the public key is submitted, the client registers the signer in the in-memory registry and in the regular disk storage as usual.

Anyway, I agree with the advantages of your approach and I'll try to follow it. It will probably need to change more things on the `keep-ecdsa` side but we'll see.

Reply:
@Shadowfiend I've tried to follow your idea and make snapshots implicitly but it seems we'll still need the possibility of doing snapshots explicitly. This is because of how our storage layer works. In the case of `keep-ecda`, if we want to snapshot the signer as soon as possible we have to do it just after the signer is generated and before the public key is published on-chain as I mentioned in the previous comment. If we make snapshots implicitly along with the regular save, we will be forced to put the signer in the `current` directory before their key validity is confirmed by the chain. Then we will need to check again and optionally remove the signer from the `current` directory (our storage api doesn't allow removing) if the `submitPublicKey` transaction failed. Also, having the signer in the `current` directory means that we will also have them in the in-memory keep registry which in turn means the client consider them as fully operating which may not be true if the `submitPublicKey` transaction has not been accepted yet. Last but not least, if we want to snapshot signatures in the future, it will be also easier to do it explicitly because we will not save them in the same way as we currently save keeps. What do you thing?

Reply:
> Before submitting the public key on-chain, the client makes a signer snapshot but doesn't persist the signer in the regular storage yet. This is because the signer is still a "draft" until the public key gets submitted on-chain so the client can't put it in the regular storage.

This is interesting. I don't think handling new write points was an explicit goal in my head, to be honest—it was just to make sure that any write that was made had a separate record of it… I think this handles the immediate need well enough that we should move it through. I think the next step is going to be an RFC to better define what we're looking for, because that seems to be a step we missed here and led to this crossing of wires.

Reply:
We may want to revisit the moment of writing shares to `current`. I don't think the current approach in persistence layer is wrong as it could protect us against overwriting key shares if not the chain reorganization but if we do a wider refactor and do append-only to `current`, reading only the first stored membership on client start, we can have all of it without `snapshots`.

On the other hand, the solution presented here is a little more complex but gives us better flexibility.